### ISSUE TITLE
validate_assignment decorator improvements

### ISSUE THREAD

**Issue by samuelcolvin** [pydantic/pydantic#1205]
Future improvements to the `validate_assignment` decorator #1179:

* [ ] arguments to the decorator, including: validators, custom config (partially fixed by #1663), return value validation
* [ ] rewrite front page to explain the 3 or 4 primary interfaces to pydantic:
  * `BaseModel`
  * `dataclasses`
  * `parse_obj_as`
  * `validate_arguments`
* [ ] perhaps change the error raised for invalid arguments to inherit from `TypeError` as other argument errors do. (Currently the normal `ValidationError` is raised which inherits from `ValueError`)
* [ ] allow `validate_assignment` to be set in "strict" mode so types are not coerced, this is really just #1098 
* [ ] option to enable or disable validation via an environment variable or similar - e.g. to only use the function in development or testing or on a subset of calls
* [ ] option to still call the function if validation fails
* [ ] get rid of the extra arguments to the underlying model so it's useful
* [ ] instance methods #1222

**Comment by RileyMShea** [comment]
Python Version: 3.8
Pydantic Version: 1.17
Vscode: v1.50.1
Pylance: v2020.10.2


Using the decorator on v1.17 with arguments seems to mangle the function signature in vscode, specifically with their pylance language server.  Haven't tested `jedi` or `microsoft` language servers.
```python
@validate_arguments(config=dict(arbitrary_types_allowed=True))
# or
@validate_arguments(config=None)

# modifies signature from def foo(a:int,b:str)->str:... to def foo()->Any:
```
`@validate_arguments` by itself works as expected though

Pycharm with `pydantic` extension does show correct signature on hover though.

**Comment by gucharbon** [comment]
@RileyMShea , my workaround for this problem:

```python
from functools import wraps
from typing import Any, Callable
from pydantic import validate_arguments as _validate_arguments


def validate_arguments(func: Callable[..., Any]) -> Callable[..., Any]:
    _func = _validate_arguments(func)

    @wraps(func)
    def wrapped_func(*args, **kwargs):
        return _func(*args, **kwargs)

    return wrapped_func
```

This way signature and docstring are preserved.

**Comment by Kilo59** [comment]
The only feature so far that I wish `@validate_arguments` could provide is to `use_enum_value`.

I ran into this in the context of building a convenience interface to a web service that takes some query parameters.

Example
https://petstore.swagger.io/#/pet/findPetsByStatus
```python
import enum
from typing import List

import httpx
from pydantic import validate_arguments

class Status(str, enum.Enum):
    available = "available"
    pending = "pending"
    sold = "sold"

@validate_arguments
def find_pets_by_status(status: List[Status]) -> List[dict]:
    r = httpx.get("https://petstore.swagger.io/v2/pet/findByStatus", params={"status": status})
    print(r.request.url)
    return r.json()

print(find_pets_by_status(["pending", "sold"]))
>>>https://petstore.swagger.io/v2/pet/findByStatus?status=Status.sold&status=Status.pending
>>>[]
```

The solution is easy enough, just extract the values for the enum. Nonetheless,  I miss `use_enum_value` 😄 .
`[s.value for s in status]`

Is this perhaps something we could just make default behavior if the enum is a mixin type?

**Comment by markedwards** [comment]
I think `validate_arguments` should support an underscore arg, for args that are ignored in the body.  Right now it fails with:
```
 File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Foo
_
  extra fields not permitted (type=value_error.extra)
```

**Comment by markedwards** [comment]
My use case for `validate_arguments` is to gather any errors related to the argument validation, and return them after applying custom formatting.  So I do something like:
```
try:
    result = validate_arguments(config=dict(arbitrary_types_allowed=True))(fn)(*args, **kwargs)
except PydanticValidationError as e:
    # apply formatting to errors
```
The issue with this is it also catches any `PydanticValidationError` exceptions that occur in the body of function, and I only want to catch and format the errors related to the input.

It would be useful to be able to get the validated arguments separately from executing the function.  Is this possible already, without effectively writing my own version of `validate_arguments`?

**Comment by PrettyWood** [comment]
Hi @markedwards 
Have you seen https://pydantic-docs.helpmanual.io/usage/validation_decorator/#validate-without-calling-the-function?

**Comment by markedwards** [comment]
@PrettyWood I had not, and that seems to do the job.  Thanks!  The only minor nitpick I would mention is mypy can't understand it, but that isn't a showstopper.

**Comment by rsokl** [comment]
Note that https://github.com/samuelcolvin/pydantic/pull/1272 added support for instance-methods; the checkbox

![image](https://user-images.githubusercontent.com/29104956/136082949-e48e12b6-fb39-4b4d-9873-06cf874641a7.png)

should be checked.

**Comment by RobertCraigie** [comment]
I would also suggest adding an argument to make model creation lazy.

I'm using `validate_arguments` on a lot of auto-generated functions with a lot of recursive / nested types and this significantly increases import time and memory usage.

e.g. `@validate_arguments(lazy=True)`

**Comment by markedwards** [comment]
Just following up on my earlier question above about getting the validated args and kwargs, I see that this is possible:
```
def validate(fn: CT) -> CT:
    validator = validate_arguments()(fn)

    @wraps(fn)
    def wrapped(*args, **kwargs):
            validated =validator.validate(*args, **kwargs)
            validated_args = [getattr(validated, validator.vd.arg_mapping[I]) for i in range(len(args))]
            validated_kwargs = {k: getattr(validated, k) for k in kwargs.keys()}
```

Can someone speak to whether there is a better way to achieve this?  Better as in less hacky and likely to continue to work?

**Comment by PhillSimonds** [comment]
Hi all!

I wrote a decorator to do custom validation on fields and args when a method is called. Pydantic does custom validation of fields at object instantiation, which works for many use cases but is inflexible for others (e.g. for optional fields that are used only in some methods). Pydantic can do basic validation on arguments, but not custom validation. The logic looks something like this:

```
    @run_custom_validations(
        [
            FieldValidator(validator=associated_device_has_primary_ip, field_names=("intf_a", "intf_z")),
            FieldValidator(validator=intf_has_one_ip_address, field_names=("intf_a", "intf_z")),
            ArgumentValidator(validator=integer_is_prime, arg_strings=("count",)),
        ]
    )
    def ping(self, count: int = 5) -> dict:
```

The validator object takes a callable and the string representation of either the fields or the args. The value of each field/arg is passed into the validator function, which raises a `ValueError` in the case that it says the data is invalid. The decorator logic catches all ValueErrors, then raises an error at the end with all of the ValueErrors included.

I'm curious if anyone else has a need for something like this, and if a contribution back makes sense? I also don't know if here is the right place to post on it, so let me know if I should redirect elsewhere :).

**Comment by mohammad-alisafaee** [comment]
Hi all,

The default value for the `arbitrary_types_allowed` parameter of `validate_arguments` should be set to `True` (currently it defaults to `False`). User types are passed to functions in many cases (about 90% in my code) and it's a bit inconvenient to set it every time. Moreover, a default value of `True` shouldn't hurt.

**Comment by ErnestoLoma** [comment]
Hi all,
 I have a class method that is used to create an instance of the class. This method has the validate_arguments annotation. Everything is fine until I add a type annotation for the return type. Then I get a NameError name is not defined.  Any suggestions would be appreciated. Below is a snippet that demonstrates the problem:
 `from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, validate_arguments


class User(BaseModel):
    class Config:
        validate_assignment = True

    id: int
    name: str = "Jane Doe"
    friend: Optional[User] = None

    def add_friend(self, friend: User):
        self.friend = friend

    @classmethod
    @validate_arguments()
    def create_bob(cls, id: int) -> User:
        return cls(id=id, name="Bob")


User.update_forward_refs()
if __name__ == "__main__":
    bob = User.create_bob(1)
`

**Comment by jayaddison** [comment]
Re: `@validate_arguments`

I'm not exactly sure how to phrase this feedback yet, but would like to share some thoughts.  Perhaps this minimal-ish repro case is a start:

```
Python 3.11.2 (main, Feb 12 2023, 00:48:52) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pydantic
>>> @pydantic.validate_arguments
... def foo(a: float | int):
...     print(a)
... 
>>> foo(1)
1.0
```

**Comment by the-matt-morris** [comment]
> Re: `@validate_arguments`
> 
> I'm not exactly sure how to phrase this feedback yet, but would like to share some thoughts. Perhaps this minimal-ish repro case is a start:
> 
> ```
> Python 3.11.2 (main, Feb 12 2023, 00:48:52) [GCC 12.2.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import pydantic
> >>> @pydantic.validate_arguments
> ... def foo(a: float | int):
> ...     print(a)
> ... 
> >>> foo(1)
> 1.0
> ```

@jayaddison I think you're looking for the [smart_union](https://docs.pydantic.dev/usage/model_config/#smart-union) config option here:

```python
import pydantic

@pydantic.validate_arguments(config={"smart_union": True})
def foo(a: float | int):
    print(a)
```

```python
foo(1)
1
```

I've run into the same thing and stumbled upon `smart_union` at some point.  It does have some caveats, but works well in this use case.

**Comment by dmontagu** [comment]
@jnj16180340 `@validate_arguments` is not meant to be put on classes, I think what you'd want is just
```
from pydantic.dataclasses import dataclass

@dataclass
class TestyClass:
     ...
```

Also, it seems to work properly if you put `@validate_arguments` under `@classmethod`:

```python
    @classmethod
    @validate_arguments
    def from_dict(cls, d):
        return cls(
            stringy=d['s'],
            inty=d['i'],
            floaty=d['f']
        )
```
(It does seem unfortunate that you can't put it _outside_/"above" the `@classmethod`, maybe we can fix that in v2...)

**Comment by malikoth** [comment]
This may be scope change, but I'd love to at least hear your thinking on this. Function signatures (optionally) contain one more type annotation that `validate_arguments` does not currently touch at all, namely the return type annotation. I would love to be able to use this not only on the data coming into a function, but also on the data going out. I wrote an implementation of this before I found `validate_arguments`, and the only thing I had to adjust with regards to return type was that if the return type is None, you don't want to try to convert return values to NoneType (or `type(None)`) you just want to replace the value with `None`.
